### PR TITLE
GH-3839: Migrate MongoDB tests to Testcontainers

### LINKED ISSUE #3839: Migrate MongoDB tests to Testcontainers
Currently our integration tests rely on the MongoDB server provided by the host of the test execution and use JUnit 4 rule for condition evaluation.
We cannot have externally configured MongoDB server on CI environment anymore, so SonarQube reports pure coverage, since most of MongoDB tests are ignored.

Switch integration tests for MongoDB to Testcontainers since Docker is available in most modern environments.
This way we also will be able to migrate all those tests to JUnit 5.

See `MosquittoContainerTest` for example.

### PR DESCRIPTION
**Pull request by oxcafedead**
Closes #3839

* Create a new base interface for Mongo container-based tests
* Migrate all tests to JUnit5
* Remove obsolete classes related to JUnit4 Mongo rule
* Fix code style & readability here and there, a few Sonar issues
* Fix failing test `validateWithConfiguredPollerFlow`
* * This test was failing with Mongo error
`$and/$or/$nor must be a nonempty array`. The cause was too small
polling ratio of the update query, the consecutive read queries
failed as no 'Oleg' document was in collection after ~100 ms.

### CODE REVIEW

File: spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/MongoDbContainerTest.java [hunk 1]
```diff
@@ -40,53 +42,68 @@
...
-					com.mongodb.reactivestreams.client.MongoClients.create(
-							MongoClientSettings.builder().uuidRepresentation(UuidRepresentation.STANDARD).build()),
-					"test");
-
-	protected MongoDatabaseFactory prepareMongoFactory(String... additionalCollectionsToDrop) {
-		cleanupCollections(MONGO_DATABASE_FACTORY, additionalCollectionsToDrop);
-		return MONGO_DATABASE_FACTORY;
+@Testcontainers(disabledWithoutDocker = true)
+public interface MongoDbContainerTest {
+	GenericContainer<?> MONGO_CONTAINER = new GenericContainer<>("mongo:5.0.9")
```
Comment (artembilan) [inline spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/MongoDbContainerTest.java:68]: It's better to have every class member surrounded with blank line: increases source code readability.

File: spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/MongoDbContainerTest.java [hunk 2]
```diff
@@ -40,53 +42,68 @@
...
 
-	protected ReactiveMongoDatabaseFactory prepareReactiveMongoFactory(String... additionalCollectionsToDrop) {
-		cleanupCollections(REACTIVE_MONGO_DATABASE_FACTORY, additionalCollectionsToDrop);
-		return REACTIVE_MONGO_DATABASE_FACTORY;
+	static void prepareReactiveMongoData(ReactiveMongoDatabaseFactory mongoDatabaseFactory, String... additionalCollectionsToDrop) {
+		cleanupCollections(mongoDatabaseFactory, additionalCollectionsToDrop);
 	}
 
-	protected void cleanupCollections(ReactiveMongoDatabaseFactory mongoDbFactory,
+	static void cleanupCollections(ReactiveMongoDatabaseFactory MONGO_DATABASE_FACTORY,
```
Comment (artembilan) [inline spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/MongoDbContainerTest.java:103]: I don't think this change was intentional.
Please, bring the parameter name back.

**Review by artembilan — CHANGES_REQUESTED** [review]
In general, this is great. As usual.
I left just a couple minor concerns and some nit-picks to make it perfect.

Thank you!

File: spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/dsl/MongoDbTests.java [hunk 3]
```diff
@@ -345,22 +349,22 @@ public MessageChannel getResultChannel() {
 		}
 
 		@Bean
-		public MongoDatabaseFactory mongoDbFactory() {
+		public MongoDatabaseFactory MONGO_DATABASE_FACTORY() {
```
[the remark this answers is not in the document]
Comment (oxcafedead) [reply spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/dsl/MongoDbTests.java:352]: A bit of sloppiness in regexp replacements. Really sorry that this was present in a pull request. 🤦‍♂️

File: spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/dsl/MongoDbTests.java [hunk 4]
```diff
@@ -46,26 +48,39 @@
...
+	@BeforeAll
+	static void prepareMongoConnection() {
+		MONGO_DATABASE_FACTORY = MongoDbContainerTest.createMongoDbFactory();
+	}
+
+	static ReactiveMongoDatabaseFactory REACTIVE_MONGO_DATABASE_FACTORY;
+
+	@BeforeAll
+	static void prepareReactiveMongoConnection() {
+		REACTIVE_MONGO_DATABASE_FACTORY = MongoDbContainerTest.createReactiveMongoDbFactory();
```
[the remark this answers is not in the document]
Comment (oxcafedead) [reply spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/dsl/MongoDbTests.java]: Sure, this makes sense

File: spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/MongoDbContainerTest.java [hunk 5]
```diff
@@ -40,53 +42,68 @@
...
+								.uuidRepresentation(UuidRepresentation.STANDARD).build()),
+				"test");
+	}
+
+	static ReactiveMongoDatabaseFactory createReactiveMongoDbFactory() {
+		return new SimpleReactiveMongoDatabaseFactory(
+				com.mongodb.reactivestreams.client.MongoClients.create(
+						MongoClientSettings.builder().applyConnectionString(new ConnectionString(
+										"mongodb://localhost:" + MONGO_CONTAINER.getFirstMappedPort()))
+								.uuidRepresentation(UuidRepresentation.STANDARD).build()),
```
[the remark this answers is not in the document]
Comment (oxcafedead) [reply spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/MongoDbContainerTest.java]: Extracted in the private method (constant could not know about mapped port, plus its visibility...)

File: spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/MongoDbContainerTest.java (hunk 2 above)
Comment (oxcafedead) [reply spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/MongoDbContainerTest.java]: Shame..

**Review by artembilan — APPROVED** [review]
LGTM.

Pulling locally for the final review and merge.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/50769d0350d223bb767fe4643ab4470817a42584

@oxcafedead , thank you again!

**Pull request closed without merging**